doc/gnpsi/gnpsi_hld.md

```diff
@@ -0,0 +1,523 @@
+
+
+### 6. High-Level Design
+
+
+#### GNPSI application
+
+Sonic runs sflow in a docker container by running an agent `hsflowd`. When sampled packets comes in, hsflowd would accumulate the packets into UDP datagrams and send them to the collector.
+
+![Current sflow stack](images/image1.png "Current sflow stack")
+
+
+After the introduction of gNPSI, a new gNPSI server process would be brought up within the sFlow container. `hsflowd` would configure an internal local loopback collector just like a sFlow collector. 
+
+**Advantage**: This local loopback approach ensures a streamlined architecture and simplifies the maintenance of the upstream code. Refer to the [Alternative](?tab=t.0#heading=h.5bgbtjgtpqgx) section on the rationale behind choosing this design.
+
+When sampled packets come in, hsflowd would send it to the local loopback collector. Then gNPSI process would read from local loopback and relay samples to the subscribed clients. 
+
+
+![sFlow switch stack with gNPSI](images/image2.png "new switch stack")
+
+
+![sFlow sample path with gNPSI](images/image3.png "sample path")
+
+
+#### SONiC gNPSI implementation
+
+The gNPSI relay server has been implemented within the [open-source gNPSI repository](https://github.com/openconfig/gnpsi) as a library. The SONiC gNPSI server integrates this relay server implementation, along with the addition of configuration, telemetry, and security functionalities.
+
+Upon initialization, the gNPSI process will retrieve path configuration from the CONFIG_DB and activate or deactivate the relay server accordingly. Furthermore, it will collect statistics from the relay server and update the COUNTERS_DB every 30 seconds.
+
+The gNPSI process has a scalability limitation of 3 clients. This modification is platform-independent.
+
+![gNPSI process sequence diagram](images/image4.png "gNPSI process sequence diagram")
+
+
+
+#### Repository and Module
+
+This design introduces a SONiC application extension. The primary modification involves adding a new `sonic-gnpsi` repository. The `sonic-gnpsi` would utilize an [open source gNPSI repo on github](https://github.com/openconfig/gnpsi) to build the server code. The `sonic-gnpsi` repo would be added as a submodule into the `sonic-buildimage`
+
+By this way we ensure modularity and maintainability, making it easier to update and manage dependencies.
+
+Build dependencies include the new `sonic-gnpsi` repository, the github `openconfig/gnpsi` repository, and `bazel`.
+
+The dependency of repository is shown below
+
+![Image showing how gNPSI is integrated into SONiC repo](images/image5.png "gnpsi repos")
+
+
+
+#### Files
+
+
+
+*   `gnpsi.cc` : main file 
+*   `/db_monitor`: to monitor config db gnpsi-related events
+*   `/server`: helper to start and stop relay server and the stats thread, also write to appl_state_db
+*   `/utils`: stats util, credential util and authz logger util.
+
+#### Feature enablement
```

Comment:
Updated a section for the default enablement of gNPSI program. This can now be controlled by FEATURE table and we can disable gNPSI program if not needed.